_posts/2023-03-20-cyclonedx-maven-aggregate-bom-why-not-to-trust.md

```diff
@@ -0,0 +1,376 @@
+  <publisher>OWASP</publisher>
+  <group>org.owasp.webgoat.lesson</group>
+  <name>auth-bypass</name>
+  <version>v8.0.0.M15</version>
+  <description>Parent Pom for the WebGoat Project. A deliberately insecure Web Application</description>
+  <purl>pkg:maven/org.owasp.webgoat.lesson/auth-bypass@v8.0.0.M15?type=jar</purl>
+  ...
+</component>
+```
+
+It is this change which, while not invalid according to the specification schema, will likely cause problems for tools which have made the assumption the component can only exist once.
+
+# Possible Future Direction in the Specification
+
+While trying to work out the best approach for solving this problem, Steve Springett, Hervé Boutemy and I spent time on a video call discussing whether the [CycloneDX maven plugin PR](https://github.com/CycloneDX/cyclonedx-maven-plugin/pull/306) I had submitted was the right approach and, if not, could we think of other possible solutions.  Part of this discussion involved deficiencies in the specification and how the specification could be updated to allow support for alternate dependency hierarchies.
+
+The problem within the specification really comes down to the fact the references used within the `dependencies` section are component references, and as such set up a 1-1 relationship between the dependency hierarchy and the component.  For example,
+
+```
+<components>
+  <component type="library" bom-ref="component1-bom-ref">
+  ...
+  </component>
+  <component type="library" bom-ref="component2-bom-ref">
+  ...
+  </component>
+</components>
+<dependencies>
+  <dependency ref="component1-bom-ref">
+    <dependency ref="component2-bom-ref"/>
+  </dependency>
+  <dependency ref="component2-bom-ref"/>
+</dependencies>
+```
+The above structure does not allow for a second top level dependency element to reference the same component since the `ref` attribute is the key to retrieve the dependency information.
+
+A solution for this would be a change to the specification to define references in the dependency section to be uniquely scoped to within that section, distinct from those used within the `component` section, and require a second reference to the `component`.  For example,
+
+```
+<components>
+  <component type="library" bom-ref="component1-bom-ref">
+  ...
+  </component>
+  <component type="library" bom-ref="component2-bom-ref">
+  ...
+  </component>
+  <component type="library" bom-ref="component3-bom-ref">
+  ...
+  </component>
+</components>
+<dependencies>
+  <dependency ref="dependency1-ref" component-ref="component1-bom-ref">
+    <dependency ref="dependency2-ref"/>
+  </dependency>
+  <dependency ref="dependency2-ref" component-ref="component2-bom-ref"/>
+
+  <dependency ref="dependency3-ref" component-ref="component1-bom-ref">
+    <dependency ref="dependency4-ref"/>
+  </dependency>
+  <dependency ref="dependency4-ref" component-ref="component3-bom-ref"/>
+</dependencies>
+```
```

Comment:
Instead of introducing a new attribute, I think the following would work:

```xml
<components>
  <component type="library" bom-ref="component1-bom-ref">
  ...
  </component>
  <component type="library" bom-ref="component2-bom-ref">
  ...
  </component>
  <component type="library" bom-ref="component3-bom-ref">
  ...
  </component>
</components>
<dependencies>
  <dependency bom-ref="dependency1-ref" ref="component1-bom-ref">
    <dependency ref="dependency2-ref"/>
  </dependency>
  <dependency bom-ref="dependency2-ref" ref="component2-bom-ref"/>
  <dependency bom-ref="dependency3-ref" ref="component1-bom-ref">
    <dependency ref="dependency4-ref"/>
  </dependency>
  <dependency bom-ref="dependency4-ref" ref="component3-bom-ref"/>
</dependencies>
```

I think this accomplishes the same thing AND because the dependency has its own bom-ref, we can use [BOM-Link](https://cyclonedx.org/capabilities/bomlink/) to reference an individual dependency from other BOMs and systems.

Reply:
I like the ides of reusing the `bom-ref` attribute, however I suspect the double meaning of `ref` within the dependency element would be confusing.  In the top level it is referencing the component and in the nested case it is referencing the top level dependency.

I was reusing `ref` to try and align with the current specs, and as we discussed on Slack the `component-ref` (or whichever name was chosen) could be an optional attribute to allow the current uses to remain supported.

Would the `bom-ref` attribute be for the 2.0 spec?

Reply:
@stevespringett I'm going to remove the specification part and publish, since that was just a suggestion.  My intention with this blog was to highlight the issues in the current implementation of the aggregate SBOM rather than advocating for specific changes to the specification.